EIPS/eip-1559.md

```diff
@@ -34,48 +38,168 @@ An important aspect of this fee system is that miners only get to keep the gas p
+		parent_gas_used = self.parent(block).gas_used
+		parent_gas_target = self.parent(block).gas_target
+		transactions_all = self.transactions(block)
+
+		# check if the block changed the gas target too much
+		assert block.gas_target <= parent_gas_target + parent_gas_target // 1024, 'invalid block: gas target increased too much'
+		assert block.gas_target >= parent_gas_target - parent_gas_target // 1024, 'invalid block: gas target decreased too much'
+
+		# check if the base fee is correct
+		excess_gas_used = parent_gas_used - parent_gas_target
+		expected_base_fee = parent_base_fee + parent_base_fee * excess_gas_used // parent_gas_target // BASE_FEE_MAX_CHANGE_DENOMINATOR
+		assert expected_base_fee == block.base_fee, 'invalid block: base fee not correct'
+
+		# execute transactions and do gas accounting
+		cumulative_legacy_transaction_gas_used = 0
+		cumulative_1559_transaction_gas_used = 0
+		for transaction in transactions_all:
+			signer = self.transaction_signer_account(transaction)
+			signer.balance -= transaction.amount
+			assert signer.balance >= 0, 'invalid transaction: signer does not have enough ETH to cover attached value'
+
+			# handle legacy transactions, can differentiate by number of items if desired (Legacy == 8 items)
+			if isinstance(transaction, TransactionLegacy):
+				signer.balance -= transaction.gas_limit * transaction.gas_price
+				assert signer.balance >= 0, 'invalid transaction: signer does not have enough ETH to cover gas'
+				gas_used = self.execute_transaction(transaction)
+				cumulative_legacy_transaction_gas_used += gas_used
+				signer.balance += (transaction.gas_limit - gas_used) * transaction.gas_price
+				self.account(block.author).balance += gas_used * transaction.gas_price
+
+			# handle 1559 transactions, can differentiate by number of items if desired (1559 == 9 items)
+			elif isinstance(transaction, Transaction1559):
+				# premium is capped such that base fee is filled first
+				premium_per_gas = min(transaction.gas_premium, transaction.fee_cap - block.base_fee)
+				# signer pays both the premium and the base fee
+				signer.balance -= transaction.gas_limit * premium_per_gas
+				signer.balance -= transaction.gas_limit * block.base_fee
+				assert signer.balance >= 0, 'invalid transaction: signer does not have enough ETH to cover gas'
+				gas_used = self.execute_transaction(transaction)
+				gas_refund = transaction.gas_limit - gas_used
+				cumulative_1559_transaction_gas_used += gas_used
+				# signer gets refunded for unused gas
+				signer.balance += gas_refund * block.base_fee
+				signer.balance += gas_refund * premium_per_gas
+				# miner only receives the premium; note that the base fee is not given to anyone (it is burned)
+				self.account(block.author).balance += gas_used * premium_per_gas
+				# if the fee cap is lower than the base fee, the miner can choose to pay for the transaction which could result in miner balance going negative
```

Comment:
This needs to be changed such that we assert `gas_price >= base_fee` and there is no deducting of coinbase balance allowed.

Reply:
I have been trying to avoid normative changes when doing large revisions like this, but since it is difficult to get engagement on these PRs I'll start bundle them unless others complain.